File: 141-linked-list-cycle/141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null || head.next == null || head.next.next == null){
            return false;
        }
        ListNode first = head;
        ListNode second = head.next;
        while(first != second){
            if(second != null){
                first = first.next;
                second = second.next;
                second = second != null ? second.next: null;
            }else{
                return false;
            }
        }
        return true;
    }
}

File: 141-linked-list-cycle/141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None or head.next == None or head.next.next == None:
            return False
        
        first = head
        second = head.next
        while first != second:
            if second != None:
                second = second.next
                first = first.next
                second = second.next if second != None else None
            else:
                return False
        
        return True